packages/compiler/core/checker.ts

```diff
@@ -1789,6 +1982,1232 @@ export function createChecker(program: Program): Checker {
+        current.push(node);
+        type.nodeByType.set(projected, node);
+        break;
+    }
+
+    return links.declaredType;
+  }
+
+  function evalProjectionNode(
+    node: ProjectionExpression | ProjectionStatementItem
+  ): TypeOrReturnRecord {
+    switch (node.kind) {
+      case SyntaxKind.ProjectionExpressionStatement:
+        return evalProjectionExpressionStatement(node);
+      case SyntaxKind.ProjectionCallExpression:
+        return evalProjectionCallExpression(node);
+      case SyntaxKind.ProjectionMemberExpression:
+        return evalProjectionMemberExpression(node);
+      case SyntaxKind.Identifier:
+        return evalProjectionIdentifier(node);
+      case SyntaxKind.ProjectionLambdaExpression:
+        return evalProjectionLambdaExpression(node);
+      case SyntaxKind.StringLiteral:
+        return evalStringLiteral(node);
+      case SyntaxKind.NumericLiteral:
+        return evalNumericLiteral(node);
+      case SyntaxKind.BooleanLiteral:
+        return evalBooleanLiteral(node);
+      case SyntaxKind.ProjectionBlockExpression:
+        return evalProjectionBlockExpression(node);
+      case SyntaxKind.ProjectionArithmeticExpression:
+        return evalProjectionArithmeticExpression(node);
+      case SyntaxKind.ProjectionIfExpression:
+        return evalProjectionIfExpression(node);
+      case SyntaxKind.ProjectionEqualityExpression:
+        return evalProjectionEqualityExpression(node);
+      case SyntaxKind.ProjectionUnaryExpression:
+        return evalProjectionUnaryExpression(node);
+      case SyntaxKind.ProjectionRelationalExpression:
+        return evalProjectionRelationalExpression(node);
+      case SyntaxKind.ProjectionReference:
+        return evalProjectionReferenceExpression(node);
+      case SyntaxKind.VoidKeyword:
+        return voidType;
+      case SyntaxKind.NeverKeyword:
+        return neverType;
+      case SyntaxKind.Return:
+        return evalReturnKeyword(node);
+      default:
+        throw new Error("Can't eval " + SyntaxKind[node.kind]);
+    }
+  }
+
+  interface EvalContext {
+    node: Node;
+    locals: Map<string, Type>;
+    parent?: EvalContext;
+  }
+
+  function evalReturnKeyword(node: ReturnExpressionNode): ReturnRecord {
+    const value = evalProjectionNode(node.value);
+    if (value.kind === "Return") {
+      return value;
+    }
+
+    return {
+      kind: "Return",
+      value,
+    };
+  }
+
+  function evalProjectionReferenceExpression(
+    node: ProjectionReferenceExpressionNode
+  ): TypeOrReturnRecord {
+    const args: Type[] = [];
+    for (const arg of node.arguments) {
+      const argVal = evalProjectionNode(arg);
+      if (argVal.kind === "Return") {
+        return argVal;
+      }
+      args.push(argVal);
+    }
+
+    const target = evalProjectionNode(node.target);
+    const projection = evalProjectionNode(node.reference);
+
+    if (projection.kind !== "Projection") {
+      throw new Error(`Can't project using a non-projection type`);
```

Comment:
I'm wondering about all the exceptions when evaluating a projection. Should these be diagnostics? I think they would have to be to give a good experience in the IDE, but also it's breaking my brain to think about the equivalent of a runtime error interpreting the projection code into a diagnostic.

Reply:
I explained this some here: https://github.com/microsoft/cadl/pull/16#discussion_r752785082. Essentially I think errors while evaluating projection code are not recoverable errors in the least. When you encounter an error, you need to stop evaluation and report the error. Exceptions are quite natural for this!

Like decorators, I think we'll have to catch this error somewhere and emit a diagnostic in cases where we want to continue checking the program or evaluating a different projection.

Reply:
I need to understand where the catch goes and how the user experience will be when one of these is hit for the user to locate and fix the problem. I'm wondering things like should we be keeping track of where in the code we did something like "negate a non-boolean".  Of course, the line of code with that negation isn't necessarily the problem, it could be that someone passed in a non-boolean where a boolean was expected. Then you almost want the interpreter stack trace. Yikes. 

I'm not sure the comparison to handling decorator throws is fair. In the decorator case, we've been saying that decorators should not throw. We'll keep the language server running and continue checking, but the experience is still basically supposed to be: "the decorator has a bug, here's a stack trace". Whereas here we're taking errors in the input program and representing them as exceptions. That would be like a decorator having an explicit throw instead of reportDiagnostic for a user error in a decorator argument, which we've been saying they should not do.

Reply:
I thought I remembered that we were going to have syntax for referencing a projection like: `model Foo { bar: Bar#v(2) }` or something. Do we plan to have that or am I just imagining that? If we ever did have that, suppose `Bar#v.to` throws on argument 2, why would we abort checking altogether on that error but not `model Foo { bar: ModelThatDoesNotExist; }`.  From there, even without that exact feature, it seems logical to think about a failed projection as an error type to me.

Reply:
Note that the language server is a singleton across all documents opened in a VSCode session, so that the main concern we addressed with exceptions in decorators is not allowing any to be unhandled, or there is no language server for the rest of the session.

In this case, I think I agree with Nick that at some point, debugging projections is part of what we want to do in the language server, and that will mean some kind of first class experience for viewing the projection exceptions. but we don't want to throw exceptions from the language server when projections throw.  For now, wrapping the exceptions in an error diagnostic seems like the right intermediate solution.

Reply:
Debugging experience will certainly need to be worked on, but we can have a good experience with exception throwing. I am not proposing that we propagate the JS exception to the user in any way - it's a device used inside the interpreter to unwind the  interpreter stack and give calling code an indication that something bad has happened. The comparison to decorators is only to say, when an exception happens, we do something that isn't just throwing the exception to the user and tearing down cadl. Agree that the something is different!

I am also not proposing that a projection reference like Foo#proj where proj throws an exception stops checking. Again, like decorators, the exception is handled and checking continues. Probably we just ignore the projection and continue checking with the base value.

Maybe useful to think what would happen if we didn't use exceptions - we would need something like an error completion result, like the return completion, and check for that  when interpreting every syntax node, and propagate it up. And in the end, I think the outcome would be identical to `throw` (interpreter stack unwound to where the projection call was made, holding some kind of record of the error that just occurred), just with a lot more code.

Reply:
Sure, in the end, the user experience is what matters. I would like to see enough code actually handle the exceptions with a reasonable user-facing diagnostic. 

> I am not proposing that we propagate the JS exception to the user in any way.

I know you aren't, but in my experience what happens when projects decide they will turn (certain) exceptions into user errors is that they often do that way too liberally and start turning program bugs into user errors. So I'd like to see the handling of these errors implemented sooner than later and understand how we're going to avoid that trap.

Reply:
I'll implement projection references to show how this could work. I think the way I'd do it is in checkProjectionReference I would try catch the call to evalProjection and if it throws, emit a diagnostic saying calling the projection failed and return the error type. In time this diagnostic can also include information about where the projection failed with a (cadl, not js) call stack, to help with debugging.

Reply:
Will the expectation be that something like this is supposed to have try/catch around checker.project call? 

 https://github.com/microsoft/cadl/pull/16/files#diff-4bd1bb3341678026ea12e18757a3631f1c0f3458023775995aa78261df388ae0R296

Or in this case, the emitter fully expects that the projection it is invoking is something it controls and it would be a bug if project threw here?

Reply:
I'm trying to understand if we actually need another feature to see an experience problem or not.

Reply:
> Or in this case, the emitter fully expects that the projection it is invoking is something it controls and it would be a bug if project threw here?

It depends. It could handle it, but how to handle it is probably emitter specific. It could also choose to trust that the projection won't throw - e.g. the versioning projections shouldn't ever throw, unless they're called wrong (e.g. passing the wrong type to the projection).

Reply:
This also feels like a bit of an aside to me. If I didn't throw exceptions inside the interpreter and instead unwound with some sort of error completion record, don't all the same problems exist? Instead of catching an exception, `checkProjectionReference` checks for an error completion, and emitters check for that too. I'm not sure that's a better experience worth more code complexity?

Reply:
I'm not necessarily asking to eliminate exceptions, I'm trying to understand the end-to-end experience and maintenance implications. It could be that exceptions remain and we get all the experiences we want. But it's really hard for me to see that with just a bunch of new `throw new Error` statements and no good examples of reporting up to the user. It took me a ton of effort to unwind (no pun intended) that from the rest of the compiler. 

That said, the difference between some sort of error return and exceptions, might be that callers would see from the typing that certain things can fail even without a bug. If emitters are supposed to do try { project() } catch { tellUserSomething() }, are they supposed to catch ALL exceptions or let others through. My worry with exceptions is that we'll catch everything and obfuscate bugs.

Maybe we could at least do something like make these new errors have a distinct type (which would give us a place to put things like the cadl stack trace, etc. in the future). I would really prefer that we don't end up with every place that invokes the interpreter having a catch-all that ends up catching null refs and such. I would prefer if we're going to expect callers to catch that we distinguish what ought to be caught in some way.

Reply:
A specific error type makes sense, was definitely planning on doing that. I was waiting til I flesh things out a bit more to understand better if we need different classes of errors (e.g. ProjectionTypeError for e.g. adding a number and a string, ProjectionMemberNotFoundError for accessing members that don't exist, etc.)